<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Prediction of chromatin looping interactions with <em>chromloop</em> • chromloop</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">chromloop</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/chromloop.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Prediction of chromatin looping interactions with <em>chromloop</em>
</h1>
                        <h4 class="author">Jonas Ibn-Salem</h4>
            <address class="author_afil">
      Faculty of Biology, Johannes Gutenberg University of Mainz, 55128 Mainz, Germany<br><a class="author_email" href="mailto:#"></a><a href="mailto:j.ibn-salem@uni-mainz.de">j.ibn-salem@uni-mainz.de</a>
      </address>
                              <h4 class="author">Miguel Andrade-Navarro</h4>
            <address class="author_afil">
      Faculty of Biology, Johannes Gutenberg University of Mainz, 55128 Mainz, Germany<br>
</address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Chromatin looping is an importent feature of eukaryotic genomes and can bring regulatory sequences, such as enhancers or transcription factor binding sites, in close physical proximity of regulated target genes. Here, we provide a tool that uses protein binding signals from ChIP-seq and sequence motif information to predict chromatin looping events. Cross-linking of proteins binding close to loop anchors result in ChIP-seq signals at both anchor loci. These signals are used at CTCF motif pairs together with their distance and orientation to each other to predict whether they interact or not. The resulting chromatin loops can be used to associate enhancers or transcription factor binding sites in form of ChIP-seq peaks to regulated target genes.</p>
    </div>
    
<div class="contents">
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Gene expression is regulated by binding of transcription factors (TF) to genomic DNA. However, many binding sites are in distal regulatory region, such as enhancers, that are hundreds of kilo bases apart of genes. These regulatory regions can physically interact with promoters of regulated genes by chromatin looping interactions. These looping interaction can be measured genome wide by chromatin conformation capture techniques such as Hi-C or ChIA-PET <span class="citation">(Rao et al. 2014, <span class="citation">Tang et al. (2015)</span>)</span>. Despite many exciting insights into the three-dimensional organization of genomes, these experimental methods are not only elaborate and expansive but also require require lots of sample material (in case of ChIA-PET) or have limited resolution of 5-40 kb (in case of Hi-C).</p>
<p>In contrast, the binding sites of TFs can be detected genome-wide by ChIP-seq experiment which are available for hundreds of TFs in many cell type and conditions. However, classical analysis of ChIP-seq gives only the direct binding sites of targeted TFs (peaks) and it is not trivial to associate them to the regulated gene without chromatin looping information.</p>
<p>Therefore, we provide a computational method to predict chromatin interactions from only genomic sequence features and ChIP-seq data. The predicted looping interactions can be used to associate TF binding sites (ChIP-seq peaks) or enhancers to regulated genes and thereby improve functional downstream analysis on the level of genes.</p>
<p>In this vignette, we show how to use the R package <em><a href="http://bioconductor.org/packages/chromloop">chromloop</a></em> to predict chromatin looping interactions between CTCF motifs by using only ChIP-seq data form a single experiment. Furthermore, we show how to train the the prediction model using custom data.</p>
</div>
<div id="predict-chromatin-interaction-form-chip-seq" class="section level1">
<h1 class="hasAnchor">
<a href="#predict-chromatin-interaction-form-chip-seq" class="anchor"></a>Predict chromatin interaction form ChIP-seq</h1>
<p>Here we show how to use the package to predict chromatin looping interactions among CTCF motif locations on chromosome 22. As input only a single bigWig file is used with coverage from a STAT1 ChIP-seq experiment in human GM12878 cells<a href="#fn1" class="footnoteRef" id="fnref1"><sup>1</sup></a>.</p>
<div id="prepare-ctcf-motif-pairs" class="section level2">
<h2 class="hasAnchor">
<a href="#prepare-ctcf-motif-pairs" class="anchor"></a>Prepare CTCF motif pairs</h2>
<p>First, we need to prepare CTCF motif pairs as candidate anchors for chromatin loop interactions. We use CTCF motif hits in human chromosome 22 as provide by <em><a href="http://bioconductor.org/packages/chromloop">chromloop</a></em> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(chromloop)
<span class="kw">library</span>(InteractionSet)

<span class="co"># load provided CTCF motifs</span>
motifs &lt;-<span class="st"> </span>motif.hg19.CTCF.chr22</code></pre></div>
<p>The CTCF motif are represented as <code>GRanges</code> object from the <em><a href="http://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></em> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(motifs)</code></pre></div>
<pre><code>## [1] "GRanges"
## attr(,"package")
## [1] "GenomicRanges"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">motifs</code></pre></div>
<pre><code>## GRanges object with 696 ranges and 5 metadata columns:
##         seqnames               ranges strand |            sequence
##            &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;character&gt;
##     [1]    chr22 [16124116, 16124134]      + | CCGCCCCGAGGGGGCAGCG
##     [2]    chr22 [16201463, 16201481]      + | CCATCGCCAGGTGGCGCCC
##     [3]    chr22 [16335406, 16335424]      + | CTGGCAGCAGATGGCGGTC
##     [4]    chr22 [17398482, 17398500]      + | TGGCCACAGGGGGGCAGTC
##     [5]    chr22 [17600547, 17600565]      + | TGACCACCAGGAGGCAGCG
##     ...      ...                  ...    ... .                 ...
##   [692]    chr22 [50964271, 50964289]      - | CGCCCTGCAGGAGGCGCTC
##   [693]    chr22 [50964897, 50964915]      - | CCGCCCGCAGGGGGCGCCC
##   [694]    chr22 [51058993, 51059011]      - | TGACCAGAAGGGGGAGCCC
##   [695]    chr22 [51135224, 51135242]      - | CAACCGACAGAGGGCGCAG
##   [696]    chr22 [51158780, 51158798]      - | CGGCCCGGCGGGGGCGGCG
##            weight      Pval   ln_Pval       sig
##         &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
##     [1]      14.0   9.3e-08   -16.191     7.032
##     [2]      12.6   3.9e-07   -14.757     6.409
##     [3]      12.0   6.8e-07   -14.201     6.167
##     [4]      11.6   9.8e-07   -13.836     6.009
##     [5]      15.3   2.1e-08   -17.679     7.678
##     ...       ...       ...       ...       ...
##   [692]      13.1   2.4e-07   -15.243     6.620
##   [693]      16.3   6.0e-09   -18.932     8.222
##   [694]      12.2   5.7e-07   -14.378     6.244
##   [695]      12.1   6.3e-07   -14.278     6.201
##   [696]      14.3   6.7e-08   -16.519     7.174
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>There are 696 CTCF motif locations on chromosome 22. The genome assembly is hg19. Four metadata columns show motif match similarity as weight score, p-value, log transformed p-value and significance score. Use <code>help(motif.hg19.CTCF)</code> to get more information on the motif dataset.</p>
<p>Now we build a dataset with all pairs of CTCF motif within 1 mega base pair and annotate it with distance, motif orientation, and motif score.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/prepareCandidates.html">prepareCandidates</a></span>(motifs, <span class="dt">maxDist =</span> <span class="dv">10</span><span class="op">^</span><span class="dv">6</span>, <span class="dt">scoreColname =</span> <span class="st">"sig"</span>)

gi</code></pre></div>
<pre><code>## StrictGInteractions object with 15586 interactions and 6 metadata columns:
##           seqnames1              ranges1     seqnames2
##               &lt;Rle&gt;            &lt;IRanges&gt;         &lt;Rle&gt;
##       [1]     chr22 [16124116, 16124134] ---     chr22
##       [2]     chr22 [16124116, 16124134] ---     chr22
##       [3]     chr22 [16124116, 16124134] ---     chr22
##       [4]     chr22 [16124116, 16124134] ---     chr22
##       [5]     chr22 [16124116, 16124134] ---     chr22
##       ...       ...                  ... ...       ...
##   [15582]     chr22 [50964897, 50964915] ---     chr22
##   [15583]     chr22 [50964897, 50964915] ---     chr22
##   [15584]     chr22 [51058993, 51059011] ---     chr22
##   [15585]     chr22 [51058993, 51059011] ---     chr22
##   [15586]     chr22 [51135224, 51135242] ---     chr22
##                        ranges2 |      dist dist_log10 strandOrientation
##                      &lt;IRanges&gt; | &lt;integer&gt;  &lt;numeric&gt;       &lt;character&gt;
##       [1] [16201463, 16201481] |     77347   4.888443           forward
##       [2] [16335406, 16335424] |    211290   5.324879           forward
##       [3] [16097012, 16097030] |     27104   4.433033         divergent
##       [4] [16405612, 16405630] |    281496   5.449472        convergent
##       [5] [17083197, 17083215] |    959081   5.981855        convergent
##       ...                  ... .       ...        ...               ...
##   [15582] [51135224, 51135242] |    170327   5.231283           reverse
##   [15583] [51158780, 51158798] |    193883   5.287540           reverse
##   [15584] [51135224, 51135242] |     76231   4.882132           reverse
##   [15585] [51158780, 51158798] |     99787   4.999074           reverse
##   [15586] [51158780, 51158798] |     23556   4.372102           reverse
##             score_1   score_2 score_min
##           &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
##       [1]     7.032     6.409     6.409
##       [2]     7.032     6.167     6.167
##       [3]     7.032     6.244     6.244
##       [4]     7.032     6.721     6.721
##       [5]     7.032     6.292     6.292
##       ...       ...       ...       ...
##   [15582]     8.222     6.201     6.201
##   [15583]     8.222     7.174     7.174
##   [15584]     6.244     6.201     6.201
##   [15585]     6.244     7.174     6.244
##   [15586]     6.201     7.174     6.201
##   -------
##   regions: 696 ranges and 5 metadata columns
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>The return value <code>gi</code> of the function <code><a href="../reference/prepareCandidates.html">prepareCandidates()</a></code> is <code>GInteractoin</code> object from the <em><a href="http://bioconductor.org/packages/InteractonSet">InteractonSet</a></em> package.</p>
</div>
<div id="parse-chip-seq-signals-and-compute-similarity-at-motif-pairs" class="section level2">
<h2 class="hasAnchor">
<a href="#parse-chip-seq-signals-and-compute-similarity-at-motif-pairs" class="anchor"></a>Parse ChIP-seq signals and compute similarity at motif pairs</h2>
<p>Now we read an example bigWig file with ChIP-seq signals. For each position along the genome, the file contains the log fold-change of ChIP signals compared to input control. A sample file with data on a subset of chromosome 22 is provided in the package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># use example ChIP-seq bigWig file</span>
bigWigFile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"GM12878_Stat1.chr22_1-18000000.bigWig"</span>, 
  <span class="dt">package =</span> <span class="st">"chromloop"</span>)

<span class="co"># add ChIP-seq coverage and compute correaltion at motif pairs</span>
gi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addCor.html">addCor</a></span>(gi, bigWigFile)</code></pre></div>
</div>
<div id="predict-loops" class="section level2">
<h2 class="hasAnchor">
<a href="#predict-loops" class="anchor"></a>Predict loops</h2>
<p>Now we can predict chromatin loops using the default trained model.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">loops &lt;-<span class="st"> </span><span class="kw"><a href="../reference/predLoops.html">predLoops</a></span>(gi)

loops</code></pre></div>
<pre><code>## StrictGInteractions object with 8 interactions and 8 metadata columns:
##       seqnames1              ranges1     seqnames2              ranges2 |
##           &lt;Rle&gt;            &lt;IRanges&gt;         &lt;Rle&gt;            &lt;IRanges&gt; |
##   [1]     chr22 [17398482, 17398500] ---     chr22 [17539262, 17539280] |
##   [2]     chr22 [17600547, 17600565] ---     chr22 [17601508, 17601526] |
##   [3]     chr22 [17600547, 17600565] ---     chr22 [17601546, 17601564] |
##   [4]     chr22 [17600547, 17600565] ---     chr22 [17652202, 17652220] |
##   [5]     chr22 [17600838, 17600856] ---     chr22 [17601546, 17601564] |
##   [6]     chr22 [17601508, 17601526] ---     chr22 [17652202, 17652220] |
##   [7]     chr22 [17601508, 17601526] ---     chr22 [17652850, 17652868] |
##   [8]     chr22 [17652500, 17652518] ---     chr22 [17652850, 17652868] |
##            dist dist_log10 strandOrientation   score_1   score_2 score_min
##       &lt;integer&gt;  &lt;numeric&gt;       &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
##   [1]    140780   5.148541        convergent     6.009     6.167     6.009
##   [2]       961   2.982723           forward     7.678     7.824     7.678
##   [3]       999   2.999565        convergent     7.678     6.456     6.456
##   [4]     51655   4.713112        convergent     7.678     6.201     6.201
##   [5]       708   2.850033        convergent     6.201     6.456     6.201
##   [6]     50694   4.704957        convergent     7.824     6.201     6.201
##   [7]     51342   4.710473        convergent     7.824    10.796     7.824
##   [8]       350   2.544068        convergent     8.509    10.796     8.509
##             chip      pred
##        &lt;numeric&gt; &lt;numeric&gt;
##   [1]  0.7902839 0.4010825
##   [2]  0.4792622 0.2784833
##   [3] -0.1316063 0.1957508
##   [4]  0.7836006 0.5188280
##   [5]  0.4247492 0.3933609
##   [6]  0.3831105 0.3251221
##   [7] -0.3702664 0.2099542
##   [8]  0.8631771 0.8555726
##   -------
##   regions: 696 ranges and 6 metadata columns
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>This result in a subset of motif pairs that are predicted to interact. The interactions are annotated with ChIP-seq correlation in column <code>chip</code> and interaction probability in <code>pred</code>.</p>
</div>
<div id="write-predicted-loops-to-output-file" class="section level2">
<h2 class="hasAnchor">
<a href="#write-predicted-loops-to-output-file" class="anchor"></a>Write predicted loops to output file</h2>
<p>Since looping interactions are stored as <code>GInteraction</code> objects, they can be exported as <a href="http://bedtools.readthedocs.io/en/latest/content/general-usage.html#bedpe-format">BED-PE</a> files using functions from <em><a href="http://bioconductor.org/packages/GenomicInteractions">GenomicInteractions</a></em> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(GenomicInteractions)</code></pre></div>
<pre><code>## Warning: replacing previous import 'BiocGenerics::Position' by
## 'ggplot2::Position' when loading 'GenomicInteractions'</code></pre>
<pre><code>## Warning: replacing previous import 'BiocGenerics::sd' by 'stats::sd' when
## loading 'GenomicInteractions'</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># export to output file</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/GenomicInteractions/topics/export.bedpe">export.bedpe</a></span>(loops, <span class="st">"loop_interactions.bedpe"</span>, <span class="dt">score =</span> <span class="st">"pred"</span>)</code></pre></div>
</div>
</div>
<div id="train-prediciton-model-using-custom-data" class="section level1">
<h1 class="hasAnchor">
<a href="#train-prediciton-model-using-custom-data" class="anchor"></a>Train prediciton model using custom data</h1>
<p>Instructions on how to fit a logistic regression model with user provided data will be explained here soon.</p>
<!-- ## Figures -->
<!-- The figure sizes have been customised so that you can easily put two images side-by-side.  -->
<!-- ```{r, fig.show='hold'} -->
<!-- plot(1:10) -->
<!-- plot(10:1) -->
<!-- ``` -->
<!-- You can enable figure captions by `fig_caption: yes` in YAML: -->
<!--     output: -->
<!--       rmarkdown::html_vignette: -->
<!--         fig_caption: yes -->
<!-- Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**. -->
<!-- ## More Examples -->
<!-- You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`. -->
<!-- ```{r, echo=FALSE, results='asis'} -->
<!-- knitr::kable(head(mtcars, 10)) -->
<!-- ``` -->
<!-- Also a quote using `>`: -->
<!-- > "He who gives up [code] safety for [code] speed deserves neither." -->
<!-- ([via](https://twitter.com/hadleywickham/status/504368538874703872)) -->
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Dunham2012">
<p>Dunham, Ian, Anshul Kundaje, Shelley F Aldred, Patrick J Collins, Carrie a Davis, Francis Doyle, Charles B Epstein, et al. 2012. “An integrated encyclopedia of DNA elements in the human genome.” <em>Nature</em> 489 (7414): 57–74. doi:<a href="https://doi.org/10.1038/nature11247">10.1038/nature11247</a>.</p>
</div>
<div id="ref-Rao2014">
<p>Rao, Suhas S P, Miriam H Huntley, Neva C Durand, Elena K Stamenova, Ivan D Bochkov, James T Robinson, Adrian L Sanborn, et al. 2014. “A 3D map of the human genome at kilobase resolution reveals principles of chromatin looping.” <em>Cell</em> 159 (7). Elsevier Inc.: 1665–80. doi:<a href="https://doi.org/10.1016/j.cell.2014.11.021">10.1016/j.cell.2014.11.021</a>.</p>
</div>
<div id="ref-Tang2015a">
<p>Tang, Zhonghui, Oscar Junhong Luo, Xingwang Li, Meizhen Zheng, Jacqueline Jufen Zhu, Przemyslaw Szalaj, Pawel Trzaskoma, et al. 2015. “CTCF-Mediated Human 3D Genome Architecture Reveals Chromatin Topology for Transcription.” <em>Cell</em>. Elsevier Inc., 1–17. doi:<a href="https://doi.org/10.1016/j.cell.2015.11.024">10.1016/j.cell.2015.11.024</a>.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>An example file with only data on a subset of chromosome 22 is provide as part of the <em><a href="http://bioconductor.org/packages/chromloop">chromloop</a></em> package. The full file can be downloaded from ENCODE <span class="citation">(Dunham et al. 2012)</span> <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeSydhTfbs/wgEncodeSydhTfbsGm12878Stat1StdSig.bigWig">here</a>.<a href="#fnref1">↩</a></p></li>
</ol>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#predict-chromatin-interaction-form-chip-seq">Predict chromatin interaction form ChIP-seq</a><ul class="nav nav-pills nav-stacked">
<li><a href="#prepare-ctcf-motif-pairs">Prepare CTCF motif pairs</a></li>
      <li><a href="#parse-chip-seq-signals-and-compute-similarity-at-motif-pairs">Parse ChIP-seq signals and compute similarity at motif pairs</a></li>
      <li><a href="#predict-loops">Predict loops</a></li>
      <li><a href="#write-predicted-loops-to-output-file">Write predicted loops to output file</a></li>
      </ul>
</li>
      <li><a href="#train-prediciton-model-using-custom-data">Train prediciton model using custom data</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Jonas Ibn-Salem.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
