<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Prediction of chromatin looping interactions with <em>sevenC</em> • sevenC</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">sevenC</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/sevenC.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ibn-salem/sevenC">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Prediction of chromatin looping interactions with <em>sevenC</em>
</h1>
                        <h4 class="author">Jonas Ibn-Salem</h4>
            <address class="author_afil">
      Faculty of Biology, Johannes Gutenberg University of Mainz, 55128 Mainz, Germany<br><a class="author_email" href="mailto:#"></a><a href="mailto:j.ibn-salem@uni-mainz.de">j.ibn-salem@uni-mainz.de</a>
      </address>
                              <h4 class="author">Miguel Andrade-Navarro</h4>
            <address class="author_afil">
      Faculty of Biology, Johannes Gutenberg University of Mainz, 55128 Mainz, Germany<br>
</address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Chromatin looping is an important feature of eukaryotic genomes and can bring regulatory sequences, such as enhancers or transcription factor binding sites, in close physical proximity of regulated target genes. Here, we provide sevenC, a tool that uses protein binding signals from ChIP-seq and sequence motif information to predict chromatin looping events. Cross-linking of proteins that bind close to loop anchors result in ChIP-seq signals at both anchor loci. These signals are used at CTCF motif pairs together with their distance and orientation to each other to predict whether they interact or not. The resulting chromatin loops can be used to associate enhancers or transcription factor binding sites (e.g. ChIP-seq peaks) to regulated target genes.</p>
    </div>
    
<div class="contents">
<div id="background-and-introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#background-and-introduction" class="anchor"></a>Background and introduction</h1>
<p>Gene expression is regulated by binding of transcription factors (TF) to genomic DNA. However, many binding sites are in distal regulatory region, such as enhancers, that are hundreds of kilo bases apart of genes. These regulatory regions can physically interact with promoters of regulated genes by chromatin looping interactions. These looping interaction can be measured genome wide by chromatin conformation capture techniques such as Hi-C or ChIA-PET <span class="citation">(Rao et al. 2014, <span class="citation">Tang et al. (2015)</span>)</span>. Despite many exciting insights into the three-dimensional organization of genomes, these experimental methods are not only elaborate and expansive but also require require lots of sample material (in case of ChIA-PET) or have limited resolution of 5-40 kb (in case of Hi-C).</p>
<p>In contrast, the binding sites of TFs can be detected genome-wide by ChIP-seq experiment which are available for hundreds of TFs in many cell type and conditions. However, classical analysis of ChIP-seq gives only the direct binding sites of targeted TFs (peaks) and it is not trivial to associate them to the regulated gene without chromatin looping information.</p>
<p>Therefore, we provide a computational method to predict chromatin interactions from only genomic sequence features and ChIP-seq data. The predicted looping interactions can be used to associate TF binding sites (ChIP-seq peaks) or enhancers to regulated genes and thereby improve functional downstream analysis on the level of genes.</p>
<p>In this vignette, we show how to use the R package <em><a href="http://bioconductor.org/packages/sevenC">sevenC</a></em> to predict chromatin looping interactions between CTCF motifs by using only ChIP-seq data form a single experiment. Furthermore, we show how to train the the prediction model using custom data.</p>
</div>
<div id="predict-chromatin-looping-interactions" class="section level1">
<h1 class="hasAnchor">
<a href="#predict-chromatin-looping-interactions" class="anchor"></a>Predict chromatin looping interactions</h1>
<div id="basic-usage-example" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-usage-example" class="anchor"></a>Basic usage example</h2>
<p>Here we show how to use the <em><a href="http://bioconductor.org/packages/sevenC">sevenC</a></em> package with default options to predict chromatin looping interactions among CTCF motif locations on chromosome 22. As input only a single bigWig file with example ChIP-seq data is used.</p>
<div id="get-motif-pairs" class="section level3">
<h3 class="hasAnchor">
<a href="#get-motif-pairs" class="anchor"></a>Get motif pairs</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(sevenC)

<span class="co"># load provided CTCF motifs in human genome</span>
motifs &lt;-<span class="st"> </span>motif.hg19.CTCF.chr22

<span class="co"># get motifs pairs</span>
gi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/prepareCisPairs.html">prepareCisPairs</a></span>(motifs, <span class="dt">scoreColname =</span> <span class="st">"sig"</span>)</code></pre></div>
</div>
<div id="add-chip-seq-data-and-compute-correaltion" class="section level3">
<h3 class="hasAnchor">
<a href="#add-chip-seq-data-and-compute-correaltion" class="anchor"></a>Add ChIP-seq data and compute correaltion</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># use example ChIP-seq bigWig file</span>
bigWigFile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"GM12878_Stat1.chr22_1-18000000.bigWig"</span>, 
  <span class="dt">package =</span> <span class="st">"sevenC"</span>)

<span class="co"># add ChIP-seq coverage and compute correaltion at motif pairs</span>
gi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addCor.html">addCor</a></span>(gi, bigWigFile)</code></pre></div>
</div>
<div id="predict-loops" class="section level3">
<h3 class="hasAnchor">
<a href="#predict-loops" class="anchor"></a>Predict loops</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># predict looping interactions among all motif pairs</span>
loops &lt;-<span class="st"> </span><span class="kw"><a href="../reference/predLoops.html">predLoops</a></span>(gi)</code></pre></div>
</div>
</div>
<div id="more-detailed-usage-example" class="section level2">
<h2 class="hasAnchor">
<a href="#more-detailed-usage-example" class="anchor"></a>More detailed usage example</h2>
<p>Here we show in more detail each step of the loop prediction process. Again, we want to predict chromatin looping interactions among CTCF motif locations on chromosome 22. As input only a single bigWig file is used with coverage from a STAT1 ChIP-seq experiment in human GM12878 cells.</p>
<div id="prepare-ctcf-motif-pairs" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-ctcf-motif-pairs" class="anchor"></a>Prepare CTCF motif pairs</h3>
<p>First, we need to prepare CTCF motif pairs as candidate anchors for chromatin loop interactions. We use CTCF motif hits in human chromosome 22 as provide by <em><a href="http://bioconductor.org/packages/sevenC">sevenC</a></em> package. In general, any CTCF motif pairs can be used if provided as <code>GRanges</code> object. To use the motif hit score, the motif data should contain a -log<sub>10</sub> transformed p-value describing the significance of each motif hit. Here, we use CTCF motif hits from RSAT matrix-scan on chromosome 22 as provided as <code>motif.hg19.CTCF.chr22</code> in the <em><a href="http://bioconductor.org/packages/sevenC">sevenC</a></em> package. For more information on the motif data set, see <code><a href="../reference/motif.hg19.CTCF.html">?motif.hg19.CTCF</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(sevenC)

<span class="co"># load provided CTCF motifs</span>
motifs &lt;-<span class="st"> </span>motif.hg19.CTCF.chr22</code></pre></div>
<p>The CTCF motif are represented as <code>GRanges</code> object from the <em><a href="http://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></em> package. There are 696 CTCF motif locations on chromosome 22. The genome assembly is hg19. Four metadata columns show motif match similarity as weight score, p-value, log transformed p-value and significance score. Use <code>help(motif.hg19.CTCF)</code> to get more information on the motif dataset.</p>
</div>
<div id="add-chip-seq-signals-at-motifs-sites" class="section level3">
<h3 class="hasAnchor">
<a href="#add-chip-seq-signals-at-motifs-sites" class="anchor"></a>Add ChIP-seq signals at motifs sites</h3>
<p>To predict loops we need the ChIP-seq signals at all motif sites. Therefore, we read an example bigWig file with ChIP-seq signals<a href="#fn1" class="footnoteRef" id="fnref1"><sup>1</sup></a>.</p>
<p>An example file with only data on a subset of chromosome 22 is provide as part of the <em><a href="http://bioconductor.org/packages/sevenC">sevenC</a></em> package. The full file can be downloaded from ENCODE <span class="citation">(Dunham et al. 2012)</span> <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeSydhTfbs/wgEncodeSydhTfbsGm12878Stat1StdSig.bigWig">here</a>. The file contains for each position in the genome the log-fold-change of ChIP-seq signals versus input control.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># use example ChIP-seq bigWig file</span>
bigWigFile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"GM12878_Stat1.chr22_1-18000000.bigWig"</span>, 
  <span class="dt">package =</span> <span class="st">"sevenC"</span>)</code></pre></div>
<p>We add ChIP-seq signals to all motifs in a window of 1000 bp using the function <code><a href="../reference/addCovToGR.html">addCovToGR()</a></code> as follows</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read ChIP-seq coverage </span>
motifs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addCovToGR.html">addCovToGR</a></span>(motifs, bigWigFile)</code></pre></div>
<p>This adds a new metadata column to <code>motifs</code> holding a <code>NumericList</code> with ChIP-seq signals for each motif location.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">motifs<span class="op">$</span>chip</code></pre></div>
<pre><code>## NumericList of length 696
## [["chr22"]] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ... 0 0 0 0 0 0 0 0 0 0 0 0 0 0
## [["chr22"]] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ... 4 4 4 4 4 4 4 4 4 4 4 4 4 4
## [["chr22"]] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0 0 0 0 0 0 0
## [["chr22"]] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ... 2 2 2 2 2 2 2 2 2 2 2 2 2 2
## [["chr22"]] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 ... 2 2 2 2 2 2 2 2 2 2 2 2 2 2
## [["chr22"]] 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 ... 3 3 3 3 3 3 3 3 3 3 3 3 3 3
## [["chr22"]] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 ... 2 2 2 2 2 2 2 2 2 3 3 3 3 3
## [["chr22"]] 2 2 2 2 2 2 2 2 2 2 2 2 ... 12 12 12 12 12 12 12 12 12 12 12
## [["chr22"]] 15 15 15 15 15 15 15 15 15 15 15 15 ... 8 8 8 8 8 8 8 8 8 8 8
## [["chr22"]] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0 0 0 0 0 0 0
## ...
## &lt;686 more elements&gt;</code></pre>
</div>
<div id="build-pairs-of-motifs-as-candidate-interactions" class="section level3">
<h3 class="hasAnchor">
<a href="#build-pairs-of-motifs-as-candidate-interactions" class="anchor"></a>Build pairs of motifs as candidate interactions</h3>
<p>Now we build a dataset with all pairs of CTCF motif within 1 Mb and annotate it with distance, motif orientation, and motif score.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/prepareCisPairs.html">prepareCisPairs</a></span>(motifs, <span class="dt">maxDist =</span> <span class="dv">10</span><span class="op">^</span><span class="dv">6</span>, <span class="dt">scoreColname =</span> <span class="st">"sig"</span>)

gi</code></pre></div>
<pre><code>## StrictGInteractions object with 15586 interactions and 5 metadata columns:
##           seqnames1              ranges1     seqnames2
##               &lt;Rle&gt;            &lt;IRanges&gt;         &lt;Rle&gt;
##       [1]     chr22 [16124116, 16124134] ---     chr22
##       [2]     chr22 [16124116, 16124134] ---     chr22
##       [3]     chr22 [16124116, 16124134] ---     chr22
##       [4]     chr22 [16124116, 16124134] ---     chr22
##       [5]     chr22 [16124116, 16124134] ---     chr22
##       ...       ...                  ... ...       ...
##   [15582]     chr22 [50964897, 50964915] ---     chr22
##   [15583]     chr22 [50964897, 50964915] ---     chr22
##   [15584]     chr22 [51058993, 51059011] ---     chr22
##   [15585]     chr22 [51058993, 51059011] ---     chr22
##   [15586]     chr22 [51135224, 51135242] ---     chr22
##                        ranges2 |      dist strandOrientation   score_1
##                      &lt;IRanges&gt; | &lt;integer&gt;       &lt;character&gt; &lt;numeric&gt;
##       [1] [16201463, 16201481] |     77347           forward     7.032
##       [2] [16335406, 16335424] |    211290           forward     7.032
##       [3] [16097012, 16097030] |     27104         divergent     7.032
##       [4] [16405612, 16405630] |    281496        convergent     7.032
##       [5] [17083197, 17083215] |    959081        convergent     7.032
##       ...                  ... .       ...               ...       ...
##   [15582] [51135224, 51135242] |    170327           reverse     8.222
##   [15583] [51158780, 51158798] |    193883           reverse     8.222
##   [15584] [51135224, 51135242] |     76231           reverse     6.244
##   [15585] [51158780, 51158798] |     99787           reverse     6.244
##   [15586] [51158780, 51158798] |     23556           reverse     6.201
##             score_2 score_min
##           &lt;numeric&gt; &lt;numeric&gt;
##       [1]     6.409     6.409
##       [2]     6.167     6.167
##       [3]     6.244     6.244
##       [4]     6.721     6.721
##       [5]     6.292     6.292
##       ...       ...       ...
##   [15582]     6.201     6.201
##   [15583]     7.174     7.174
##   [15584]     6.201     6.201
##   [15585]     7.174     6.244
##   [15586]     7.174     6.201
##   -------
##   regions: 696 ranges and 6 metadata columns
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>The function <code><a href="../reference/prepareCisPairs.html">prepareCisPairs()</a></code> returns a <code>GInteractoin</code> object from the <em><a href="http://bioconductor.org/packages/InteractonSet">InteractonSet</a></em> package, representing all motif pairs within the defined distance. Metadata columns of the <code>GInteractoin</code> object hold the genomic distance between motifs in bp (<code>dist</code>), the orientation of motifs (<code>strandOrientation</code>), and motif score as -log<sub>10</sub> of the motif hit p-value (<code>score_1</code>, <code>score_2</code>, and <code>score_min</code>). Note, that the function <code><a href="../reference/prepareCisPairs.html">prepareCisPairs()</a></code> is wrapper for three individual functions that performs each step separately and allow more options. First, <code><a href="../reference/getCisPairs.html">getCisPairs()</a></code> is used to builds the <code>GInteractoin</code> object. Than <code><a href="../reference/addStrandCombination.html">addStrandCombination()</a></code> adds the four possible strand combinations of motifs pairs. Finally, <code><a href="../reference/addMotifScore.html">addMotifScore()</a></code> adds the minimum motif score for each pair. These features are later used as variables for predictions.</p>
</div>
</div>
<div id="compute-chip-seq-similarity-at-motif-pairs" class="section level2">
<h2 class="hasAnchor">
<a href="#compute-chip-seq-similarity-at-motif-pairs" class="anchor"></a>Compute ChIP-seq similarity at motif pairs</h2>
<p>Now, we compute the correlation of ChIP-seq signals for all motif pairs across positions around motif centers. Essentially, for two motifs the corresponding ChIP-seq signal vectors that were added to <code>motifs</code> before, are compared by Pearson correlation. High correlation of ChIP-seq signals at two motifs indicate a similar ChIP-seq coverage profile at the two motifs. This in turn is characteristic for interacting motifs, where ChIP signals are found on both side with a similar distance to motif centers. The correlation coefficient is added as additional metadata column to <code>gi</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># add ChIP-seq coverage and compute correaltion at motif pairs</span>
gi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addCovCor.html">addCovCor</a></span>(gi)</code></pre></div>
</div>
<div id="predict-loops-1" class="section level2">
<h2 class="hasAnchor">
<a href="#predict-loops-1" class="anchor"></a>Predict loops</h2>
<p>Now we can predict chromatin loops integrating the ChIP-seq correlation with other genomic features in a logistic regression model.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">loops &lt;-<span class="st"> </span><span class="kw"><a href="../reference/predLoops.html">predLoops</a></span>(gi)

loops</code></pre></div>
<pre><code>## StrictGInteractions object with 8 interactions and 7 metadata columns:
##       seqnames1              ranges1     seqnames2              ranges2 |
##           &lt;Rle&gt;            &lt;IRanges&gt;         &lt;Rle&gt;            &lt;IRanges&gt; |
##   [1]     chr22 [17398482, 17398500] ---     chr22 [17539262, 17539280] |
##   [2]     chr22 [17600547, 17600565] ---     chr22 [17601508, 17601526] |
##   [3]     chr22 [17600547, 17600565] ---     chr22 [17601546, 17601564] |
##   [4]     chr22 [17600547, 17600565] ---     chr22 [17652202, 17652220] |
##   [5]     chr22 [17600838, 17600856] ---     chr22 [17601546, 17601564] |
##   [6]     chr22 [17601508, 17601526] ---     chr22 [17652202, 17652220] |
##   [7]     chr22 [17601508, 17601526] ---     chr22 [17652850, 17652868] |
##   [8]     chr22 [17652500, 17652518] ---     chr22 [17652850, 17652868] |
##            dist strandOrientation   score_1   score_2 score_min   cor_chip
##       &lt;integer&gt;       &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;
##   [1]    140780        convergent     6.009     6.167     6.009  0.7902839
##   [2]       961           forward     7.678     7.824     7.678  0.4792622
##   [3]       999        convergent     7.678     6.456     6.456 -0.1316063
##   [4]     51655        convergent     7.678     6.201     6.201  0.7836006
##   [5]       708        convergent     6.201     6.456     6.201  0.4247492
##   [6]     50694        convergent     7.824     6.201     6.201  0.3831105
##   [7]     51342        convergent     7.824    10.796     7.824 -0.3702664
##   [8]       350        convergent     8.509    10.796     8.509  0.8631771
##            pred
##       &lt;numeric&gt;
##   [1] 0.4010825
##   [2] 0.2784833
##   [3] 0.1957508
##   [4] 0.5188280
##   [5] 0.3933609
##   [6] 0.3251221
##   [7] 0.2099542
##   [8] 0.8555726
##   -------
##   regions: 696 ranges and 6 metadata columns
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>This result in a subset of motif pairs that are predicted to interact. The interactions are annotated with ChIP-seq correlation in column <code>chip</code>. The column <code>pred</code> holds the predicted interaction probability according to the logistic regression model.</p>
<p>Note, that without specifying further options, the function <code><a href="../reference/predLoops.html">predLoops()</a></code> uses a default model that was optimized for several transcription factor ChIP-seq data sets by using experimental chromatin loops from Hi-C and ChIA-PET for validations. However, users can specify custom features using the <code>formular</code> argument and provide custom parameters using the <code>betas</code>argument. Furthermore, per default the <code><a href="../reference/predLoops.html">predLoops()</a></code> function report only looping interactions that mead a minimal prediction score threshold. This can be modified using the <code>cutoff</code> argument.</p>
</div>
</div>
<div id="downstream-analysis-with-chromatin-loops" class="section level1">
<h1 class="hasAnchor">
<a href="#downstream-analysis-with-chromatin-loops" class="anchor"></a>Downstream analysis with chromatin loops</h1>
<div id="linking-sets-of-regions" class="section level2">
<h2 class="hasAnchor">
<a href="#linking-sets-of-regions" class="anchor"></a>Linking sets of regions</h2>
<p>Predicted loops are represented as <code>GInteraction</code> and can therefore be used easily for downstream analysis with functions from the <em><a href="http://bioconductor.org/packages/InteractonSet">InteractonSet</a></em> package. For example linking two sets of regions (like ChIP-seq peaks and genes) can be done using the <code>linkOverlaps</code> function. See the <a href="http://bioconductor.org/packages/release/bioc/vignettes/InteractionSet/inst/doc/interactions.html">vignette</a> of the <em><a href="http://bioconductor.org/packages/InteractonSet">InteractonSet</a></em> package for more details and examples.</p>
</div>
<div id="write-predicted-loops-to-output-file" class="section level2">
<h2 class="hasAnchor">
<a href="#write-predicted-loops-to-output-file" class="anchor"></a>Write predicted loops to output file</h2>
<p>Since looping interactions are stored as <code>GInteraction</code> objects, they can be exported as <a href="http://bedtools.readthedocs.io/en/latest/content/general-usage.html#bedpe-format">BED-PE</a> files using functions from <em><a href="http://bioconductor.org/packages/GenomicInteractions">GenomicInteractions</a></em> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GenomicInteractions)</code></pre></div>
<pre><code>## Warning: replacing previous import 'BiocGenerics::Position' by
## 'ggplot2::Position' when loading 'GenomicInteractions'</code></pre>
<pre><code>## Warning: replacing previous import 'BiocGenerics::sd' by 'stats::sd' when
## loading 'GenomicInteractions'</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># export to output file</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/GenomicInteractions/topics/export.bedpe">export.bedpe</a></span>(loops, <span class="st">"loop_interactions.bedpe"</span>, <span class="dt">score =</span> <span class="st">"pred"</span>)</code></pre></div>
</div>
</div>
<div id="train-prediciton-model-using-custom-data" class="section level1">
<h1 class="hasAnchor">
<a href="#train-prediciton-model-using-custom-data" class="anchor"></a>Train prediciton model using custom data</h1>
<p>Here, we show how to use <em><a href="http://bioconductor.org/packages/sevenC">sevenC</a></em> to build and train a logistic regression model for loop prediction.</p>
<div id="prepare-moitf-pairs-and-add-chip-seq-data" class="section level2">
<h2 class="hasAnchor">
<a href="#prepare-moitf-pairs-and-add-chip-seq-data" class="anchor"></a>Prepare moitf pairs and add ChIP-seq data</h2>
<p>First, we need to build the pairs of motifs as candidate and add the ChIP-seq data as shown above.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load provided CTCF motifs</span>
motifs &lt;-<span class="st"> </span>motif.hg19.CTCF.chr22

<span class="co"># use example ChIP-seq coverage file</span>
bigWigFile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"GM12878_Stat1.chr22_1-18000000.bigWig"</span>, 
  <span class="dt">package =</span> <span class="st">"sevenC"</span>)

<span class="co"># add ChIP-seq coverage</span>
motifs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addCovToGR.html">addCovToGR</a></span>(motifs, bigWigFile)

<span class="co"># build motif pairs</span>
gi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/prepareCisPairs.html">prepareCisPairs</a></span>(motifs, <span class="dt">maxDist =</span> <span class="dv">10</span><span class="op">^</span><span class="dv">6</span>, <span class="dt">scoreColname =</span> <span class="st">"sig"</span>)

<span class="co"># add correaltion of ChIP-signal</span>
gi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addCovCor.html">addCovCor</a></span>(gi)</code></pre></div>
</div>
<div id="train-predictor-with-known-loops" class="section level2">
<h2 class="hasAnchor">
<a href="#train-predictor-with-known-loops" class="anchor"></a>Train predictor with known loops</h2>
<p>We need to label true looping interactions by using experimental data of chromatin interactions. Here, we use loops from high resolution Hi-C experiments in human GM12878 cells <span class="citation">(Rao et al. 2014)</span>. An example file with loops on chromosome 22 is provided with the <em><a href="http://bioconductor.org/packages/sevenC">sevenC</a></em> package and the function <code><a href="../reference/parseLoopsRao.html">parseLoopsRao()</a></code> reads loops in the format provided by Rao et al. and returns a <code>GInteraction</code> object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># parse known loops</span>
knownLoopFile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, 
  <span class="st">"GM12878_HiCCUPS.chr22_1-18000000.loop.txt"</span>, <span class="dt">package =</span> <span class="st">"sevenC"</span>)

knownLoops &lt;-<span class="st"> </span><span class="kw"><a href="../reference/parseLoopsRao.html">parseLoopsRao</a></span>(knownLoopFile)</code></pre></div>
<p>We can add a new metadata column to the motif pairs <code>gi</code>, indicating whether the pair is interacting in the experimental data using the function <code><a href="../reference/addInteractionSupport.html">addInteractionSupport()</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># add known loops</span>
gi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addInteractionSupport.html">addInteractionSupport</a></span>(gi, knownLoops)</code></pre></div>
<p>The experimental support is added as factor with levels <code>"Loop"</code> and <code>"No loop"</code> as metadata column named <code>loop</code>. The column name can be modified using the <code>colname</code> argument.</p>
</div>
<div id="train-logistic-regression-model" class="section level2">
<h2 class="hasAnchor">
<a href="#train-logistic-regression-model" class="anchor"></a>Train logistic regression model</h2>
<p>We can use the <code>glm()</code> function in R to fit a logistic regression model in which the <code>loop</code> column is the dependent variable and the ChIP-seq correlation, distance, and strand orientation are the predictors.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit &lt;-<span class="st"> </span><span class="kw">glm</span>(
  <span class="dt">formula =</span> loop <span class="op">~</span><span class="st"> </span>cor_chip <span class="op">+</span><span class="st"> </span>dist <span class="op">+</span><span class="st"> </span>strandOrientation, 
  <span class="dt">data =</span> <span class="kw">mcols</span>(gi), 
  <span class="dt">family =</span> <span class="kw">binomial</span>()
  )</code></pre></div>
</div>
<div id="predict-loops-with-custom-model" class="section level2">
<h2 class="hasAnchor">
<a href="#predict-loops-with-custom-model" class="anchor"></a>Predict loops with custom model</h2>
<p>Now, we can use this model to add predicted looping probabilities.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># add predict loops</span>
gi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/predLoops.html">predLoops</a></span>(
  gi,
  <span class="dt">formula =</span> loop <span class="op">~</span><span class="st"> </span>cor_chip <span class="op">+</span><span class="st"> </span>dist <span class="op">+</span><span class="st"> </span>strandOrientation,
  <span class="dt">betas =</span> <span class="kw">coef</span>(fit),
  <span class="dt">cutoff =</span> <span class="ot">NULL</span>
)</code></pre></div>
<p>Here, we have to use the same formula as argument as in the model fitting step above. The <code>betas</code> argument takes the coefficients of the logistic regression model. Finally, the argument <code>cutoff = NULL</code> ensures that no filtering is done and all input candidates are reported. The prediction score is added as a new metadata column to <code>gi</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gi </code></pre></div>
<pre><code>## StrictGInteractions object with 15586 interactions and 8 metadata columns:
##           seqnames1              ranges1     seqnames2
##               &lt;Rle&gt;            &lt;IRanges&gt;         &lt;Rle&gt;
##       [1]     chr22 [16124116, 16124134] ---     chr22
##       [2]     chr22 [16124116, 16124134] ---     chr22
##       [3]     chr22 [16124116, 16124134] ---     chr22
##       [4]     chr22 [16124116, 16124134] ---     chr22
##       [5]     chr22 [16124116, 16124134] ---     chr22
##       ...       ...                  ... ...       ...
##   [15582]     chr22 [50964897, 50964915] ---     chr22
##   [15583]     chr22 [50964897, 50964915] ---     chr22
##   [15584]     chr22 [51058993, 51059011] ---     chr22
##   [15585]     chr22 [51058993, 51059011] ---     chr22
##   [15586]     chr22 [51135224, 51135242] ---     chr22
##                        ranges2 |      dist strandOrientation   score_1
##                      &lt;IRanges&gt; | &lt;integer&gt;       &lt;character&gt; &lt;numeric&gt;
##       [1] [16201463, 16201481] |     77347           forward     7.032
##       [2] [16335406, 16335424] |    211290           forward     7.032
##       [3] [16097012, 16097030] |     27104         divergent     7.032
##       [4] [16405612, 16405630] |    281496        convergent     7.032
##       [5] [17083197, 17083215] |    959081        convergent     7.032
##       ...                  ... .       ...               ...       ...
##   [15582] [51135224, 51135242] |    170327           reverse     8.222
##   [15583] [51158780, 51158798] |    193883           reverse     8.222
##   [15584] [51135224, 51135242] |     76231           reverse     6.244
##   [15585] [51158780, 51158798] |     99787           reverse     6.244
##   [15586] [51158780, 51158798] |     23556           reverse     6.201
##             score_2 score_min   cor_chip     loop         pred
##           &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;factor&gt;    &lt;numeric&gt;
##       [1]     6.409     6.409 -0.1371194  No loop 9.782302e-10
##       [2]     6.167     6.167       &lt;NA&gt;  No loop         &lt;NA&gt;
##       [3]     6.244     6.244 -0.2117460  No loop 6.638165e-10
##       [4]     6.721     6.721       &lt;NA&gt;  No loop         &lt;NA&gt;
##       [5]     6.292     6.292  0.7153757  No loop 2.363332e-01
##       ...       ...       ...        ...      ...          ...
##   [15582]     6.201     6.201       &lt;NA&gt;  No loop         &lt;NA&gt;
##   [15583]     7.174     7.174       &lt;NA&gt;  No loop         &lt;NA&gt;
##   [15584]     6.201     6.201       &lt;NA&gt;  No loop         &lt;NA&gt;
##   [15585]     7.174     6.244       &lt;NA&gt;  No loop         &lt;NA&gt;
##   [15586]     7.174     6.201       &lt;NA&gt;  No loop         &lt;NA&gt;
##   -------
##   regions: 696 ranges and 6 metadata columns
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>As a very simple validation we can now compare the prediction score for looping and non-looping motif pairs using a boxplot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">boxplot</span>(gi<span class="op">$</span>pred <span class="op">~</span><span class="st"> </span>gi<span class="op">$</span>loop, 
        <span class="dt">ylab =</span> <span class="st">"Predicted interaction probability"</span>)</code></pre></div>
<p><img src="sevenC_files/figure-html/unnamed-chunk-21-1.png" width="288"></p>
<p>This shows higher prediction scores for truly looping motif pairs. However, this is an insufficient evaluation of prediction performance, since the prediction score is evaluated on the same data as it was trained. A more detailed evaluation of prediction performance using cross-validation and different cell types is soon available in the sevenC paper.</p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Dunham2012">
<p>Dunham, Ian, Anshul Kundaje, Shelley F Aldred, Patrick J Collins, Carrie a Davis, Francis Doyle, Charles B Epstein, et al. 2012. “An integrated encyclopedia of DNA elements in the human genome.” <em>Nature</em> 489 (7414): 57–74. doi:<a href="https://doi.org/10.1038/nature11247">10.1038/nature11247</a>.</p>
</div>
<div id="ref-Rao2014">
<p>Rao, Suhas S P, Miriam H Huntley, Neva C Durand, Elena K Stamenova, Ivan D Bochkov, James T Robinson, Adrian L Sanborn, et al. 2014. “A 3D map of the human genome at kilobase resolution reveals principles of chromatin looping.” <em>Cell</em> 159 (7). Elsevier Inc.: 1665–80. doi:<a href="https://doi.org/10.1016/j.cell.2014.11.021">10.1016/j.cell.2014.11.021</a>.</p>
</div>
<div id="ref-Tang2015a">
<p>Tang, Zhonghui, Oscar Junhong Luo, Xingwang Li, Meizhen Zheng, Jacqueline Jufen Zhu, Przemyslaw Szalaj, Pawel Trzaskoma, et al. 2015. “CTCF-Mediated Human 3D Genome Architecture Reveals Chromatin Topology for Transcription.” <em>Cell</em>. Elsevier Inc., 1–17. doi:<a href="https://doi.org/10.1016/j.cell.2015.11.024">10.1016/j.cell.2015.11.024</a>.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Note, reading of bigWig files is currently not supported on Windows. See <code>help(rtracklayer::import.bw)</code> for more information. Users on Windows need to add ChIP-seq signals as <code>NumierList</code> by other means.<a href="#fnref1">↩</a></p></li>
</ol>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#background-and-introduction">Background and introduction</a></li>
      <li>
<a href="#predict-chromatin-looping-interactions">Predict chromatin looping interactions</a><ul class="nav nav-pills nav-stacked">
<li><a href="#basic-usage-example">Basic usage example</a></li>
      <li><a href="#more-detailed-usage-example">More detailed usage example</a></li>
      <li><a href="#compute-chip-seq-similarity-at-motif-pairs">Compute ChIP-seq similarity at motif pairs</a></li>
      <li><a href="#predict-loops-1">Predict loops</a></li>
      </ul>
</li>
      <li>
<a href="#downstream-analysis-with-chromatin-loops">Downstream analysis with chromatin loops</a><ul class="nav nav-pills nav-stacked">
<li><a href="#linking-sets-of-regions">Linking sets of regions</a></li>
      <li><a href="#write-predicted-loops-to-output-file">Write predicted loops to output file</a></li>
      </ul>
</li>
      <li>
<a href="#train-prediciton-model-using-custom-data">Train prediciton model using custom data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#prepare-moitf-pairs-and-add-chip-seq-data">Prepare moitf pairs and add ChIP-seq data</a></li>
      <li><a href="#train-predictor-with-known-loops">Train predictor with known loops</a></li>
      <li><a href="#train-logistic-regression-model">Train logistic regression model</a></li>
      <li><a href="#predict-loops-with-custom-model">Predict loops with custom model</a></li>
      </ul>
</li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Jonas Ibn-Salem.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
