<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Add correlation of anchor signals to pairs of close genomic regions. — addCovCor • chromloop</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">chromloop</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/chromloop.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ibn-salem/chromloop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Add correlation of anchor signals to pairs of close genomic regions.</h1>
    </div>

    
    <p>This function adds a vector with correlation values for each input
interaction. Only works for input interaction within the given <code>maxDist</code>
distance. Note, this function does not work on windows because reading of
bigWig files is currently not supported on windows.</p>
    

    <pre class="usage"><span class='fu'>addCovCor</span>(<span class='no'>gi</span>, <span class='kw'>datacol</span> <span class='kw'>=</span> <span class='st'>"chip"</span>, <span class='kw'>colname</span> <span class='kw'>=</span> <span class='st'>"cor_chip"</span>, <span class='kw'>maxDist</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gi</th>
      <td><p>A sorted <code><a href='http://www.rdocumentation.org/packages/InteractionSet/topics/GInteractions-class'>GInteractions</a></code> object.</p></td>
    </tr>
    <tr>
      <th>datacol</th>
      <td><p>a string matching an annotation column in <code>regions(gi)</code>.
This column is assumed to hold the same number of values for each
interaction as a <code>NumericList</code>.</p></td>
    </tr>
    <tr>
      <th>colname</th>
      <td><p>A string that is used as columnname for the new column in
<code>gi</code>.</p></td>
    </tr>
    <tr>
      <th>maxDist</th>
      <td><p>maximal distance of pairs in bp as numeric. If maxDist=NULL,
the maximal distance is computed from input interactions gi by
<code>max(pairdist(gi))</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code><a href='http://www.rdocumentation.org/packages/InteractionSet/topics/GInteractions-class'>GInteractions</a></code> similar to <code>gi</code>
 just with an additional column added.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='no'>.Platform</span>$<span class='no'>OS.type</span> <span class='kw'>!=</span> <span class='st'>"windows"</span>) {

  <span class='co'># use internal motif data on chromosome 22</span>
  <span class='no'>motifGR</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>chromloop</span><span class='kw ns'>::</span><span class='no'><a href='http://www.rdocumentation.org/packages/chromloop/topics/motif.hg19.CTCF.chr22'>motif.hg19.CTCF.chr22</a></span>

  <span class='co'># use example bigWig file</span>
  <span class='no'>exampleBigWig</span> <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata"</span>,
  <span class='st'>"GM12878_Stat1.chr22_1-18000000.bigWig"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"chromloop"</span>)

  <span class='co'># add coverage from bigWig file</span>
  <span class='no'>motifGR</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='addCovToGR.html'>addCovToGR</a></span>(<span class='no'>motifGR</span>, <span class='no'>exampleBigWig</span>)

  <span class='co'># get all pairs within 1Mb</span>
  <span class='no'>gi</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='getCisPairs.html'>getCisPairs</a></span>(<span class='no'>motifGR</span>, <span class='fl'>1e5</span>)

  <span class='co'># compute correaltion of coverge for each pair</span>
  <span class='no'>gi</span> <span class='kw'>&lt;-</span> <span class='fu'>addCovCor</span>(<span class='no'>gi</span>)

  <span class='co'># addCovCor adds a new metadata column:</span>
  <span class='fu'>mcols</span>(<span class='no'>gi</span>)

}</div><div class='output co'>#&gt; DataFrame with 1835 rows and 2 columns
#&gt;           dist   cor_chip
#&gt;      &lt;integer&gt;  &lt;numeric&gt;
#&gt; 1        77347 -0.1371194
#&gt; 2        27104 -0.2117460
#&gt; 3        70206         NA
#&gt; 4        25489  0.7400029
#&gt; 5          291 -0.1884298
#&gt; ...        ...        ...
#&gt; 1831     94722         NA
#&gt; 1832     94096         NA
#&gt; 1833     76231         NA
#&gt; 1834     99787         NA
#&gt; 1835     23556         NA</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Jonas Ibn-Salem.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
